desktop-widgets/importgps.cpp

```diff
@@ -0,0 +1,329 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <QDebug>
+
+#define BUFFER_SIZE 5000
+
+/* Import dive coordinates from a GPS device and synchronise them with the dive profile information
+   of a dive computer. This file contains the infrastructure to:
+   1) Read a .GPX file from a GPS system.
+   2) Find the first gpx trackpoint that follows after the start of the dive.
+   3) Allow modification of the coordinates dependent on international time zone and
+              on differences in local time settings between the GPS and the dive computer.
+   4) Saving the coordinates into the Coordinates text box in the Dive Site Edit panel
+              and which which causes the map to show the location of the dive site. 
+   The structure coords is used to store critical information.  */
+ImportGPS::ImportGPS(QWidget *parent, QString fileName, class Ui::LocationInformation *LocationUI) : QDialog(parent),
+	fileName(fileName), LocationUI(LocationUI)
+{
+	ui.setupUi(this);
+	connect(ui.timeDiffEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeDiffEditChanged()));
+	connect(ui.timeZoneEdit, SIGNAL(timeChanged(const QTime)), this, SLOT(timeZoneEditChanged()));
+	connect(ui.timezone_backwards, SIGNAL(toggled(bool)), this, SLOT(changeZoneBackwards()));
+	connect(ui.timezone_forward, SIGNAL(toggled(bool)), this, SLOT(changeZoneForward()));
+	connect(ui.diff_backwards, SIGNAL(toggled(bool)), this, SLOT(changeDiffBackwards()));
+	connect(ui.diff_forward, SIGNAL(toggled(bool)), this, SLOT(changeDiffForward()));
+	connect(ui.GPSbuttonBox, SIGNAL(clicked(QAbstractButton *)), this, SLOT(buttonClicked(QAbstractButton *)));
+	coords.settingsDiff_offset = 0;
+	coords.timeZone_offset = 0;
+	coords.lat = 0;
+	coords.lon = 0;
+	pixmapSize = (int) (ui.diveDateLabel->height() / 2);
+}
+
+void  ImportGPS::buttonClicked(QAbstractButton *button)
+{
+	if (ui.GPSbuttonBox->buttonRole(button) == QDialogButtonBox::AcceptRole) {
+		// Write the coordinates in decimal degree format to the Coordinates QLineEdit of the LocationaInformationWidget UI:
+		LocationUI->diveSiteCoordinates->setText(QString::number(coords.lat) + ", " + QString::number(coords.lon));
+		LocationUI->diveSiteCoordinates->editingFinished();
+	} else {
+		close();
+	}
+}
+
+// Extract text from the present position in the file until
+//  the first 'delim' character is encountered.
+int ImportGPS::get_substring(FILE *f, char *buf, char delim)
+{
+	int j;
+	char c;
+	for (j = 0; (c = fgetc(f)); j++) {
+		if (c == delim) break;
+		if (c == EOF) return 1;
+		buf[j] = c;
+		if (j >= BUFFER_SIZE)
+			return 2;
+	}
+	buf[j] = 0x0; // terminate the string
+	return 0;
+}
+
+// Find the next occurence of a specified GPX element in the file,
+// characerised by a "<xxx " or "<xxx>" character sequence.
+// s specifies the name of the element searched for.
+// termc is the ending character of the element name search = '>' or ' '.
+int ImportGPS::find_xml_element(FILE *f, const char *s, char *buf, char termc)
```

Comment:
instead of more complex C logic it might be better to leverage Qt's utilities such as QChar.
this is not in the C backend and performance of QChar should be pretty good.

this is not strictly a blocker, but makes the CPP code more readable instead of the C / C++ mixture within a single CPP file.

Reply:
Done. Qt constructs were used where possible.